### PR TITLE
[ENG-5782] Citation information from addons service 

### PR DESCRIPTION
**Pull request by Vlad0n20**

### CODE REVIEW

File: website/citations/views.py [hunk 1]
```diff
@@ -183,7 +196,87 @@ def citation_list(self):
         def _citation_list(auth, node_addon, list_id=None, **kwargs):
             """ Returns a list of citations
             """
+
             show = request.args.get('view', 'all')
-            return Provider().citation_list(node_addon, auth.user, list_id, show)
+            if not waffle.flag_is_active(request, features.ENABLE_GV):
+                return Provider().citation_list(node_addon, auth.user, list_id, show)
+            else:
+                response = dict(contents=[])
```
Comment (opaduchak) [inline website/citations/views.py:204]:
```suggestion
                response = {'contents': []}
```

File: website/citations/views.py [hunk 2]
```diff
@@ -183,7 +196,87 @@ def citation_list(self):
...
+                project_ulr = settings.DOMAIN + request.view_args.get('pid')
```
Comment (opaduchak) [inline website/citations/views.py:205]: TYPO:
```suggestion
                project_url = settings.DOMAIN + request.view_args.get('pid')
```

File: website/citations/views.py [hunk 3]
```diff
@@ -183,7 +196,87 @@ def citation_list(self):
...
+                project = Node.objects.filter(guids___id__in=[request.view_args.get('pid')]).first()
+                resource_references_response = get_gv_result(
+                    endpoint_url=RESOURCE_LIST_ENDPOINT,
+                    requesting_user=auth.user,
+                    requested_resource=project,
+                    params={'filter[resource_uri]': project_ulr},
```
Comment (opaduchak) [inline website/citations/views.py:211]: TYPO
```suggestion
                    params={'filter[resource_uri]': project_url},
```

File: website/citations/views.py [hunk 4]
```diff
@@ -183,7 +196,87 @@ def citation_list(self):
...
+                    params={}
+                )
+                # citation_list = list(
+                #     filter(lambda x: x['attributes']['external_service_name'] == addon_short_name, addons_url_list))
+
+                # resource_references_response = get_gv_result(
+                #     endpoint_url=USER_LIST_ENDPOINT,
+                #     requesting_user=auth.user,
+                #     params={'filter[user_uri]': f'{settings.DOMAIN}/{auth.user._id}'},
+                # )
```
Comment (opaduchak) [inline website/citations/views.py:229]: please either remove commented out code, or uncomment it

File: website/citations/views.py [hunk 5]
```diff
@@ -183,7 +196,87 @@ def citation_list(self):
...
+                for addon in addons_url_list:
+                    gv_response = _make_gv_request(
```
Comment (opaduchak) [inline website/citations/views.py:231]: Please, don't use protected attributes from other modules
I'd suggest you to implement helper function to execute `AddonOperationInvocation`s

File: website/citations/views.py [hunk 6]
```diff
@@ -183,7 +196,87 @@ def citation_list(self):
...
+                    )
+                    if gv_response.status_code == 201:
+                        attributes_dict = gv_response.json()['data']['attributes']
+                        items = attributes_dict.get('operation_result').get('items')
+                        item_id = items[0].get('item_id', '')
+
+                        change_response = {
+                            'data': {
+                                'addon': addon_short_name,
+                                'kind': items[0].get('item_type', '').lower(),
```
Comment (opaduchak) [inline website/citations/views.py:264]: Why do you return only the first item from GV response?

File: website/citations/views.py [hunk 7]
```diff
@@ -183,7 +196,87 @@ def citation_list(self):
...
+                                'id': item_id,
+                                'name': items[0].get('item_name', ''),
+                                'path': items[0].get('item_path', ''),
+                                'parent_list_id': 'ROOT',
```
Comment (opaduchak) [inline website/citations/views.py:268]: I don't think that this would be universal for every invocation

File: website/citations/views.py [hunk 8]
```diff
@@ -183,7 +196,87 @@ def citation_list(self):
...
+                )
+                configured_storage_addons_url = resource_references_response.get_related_link(
+                    'configured_storage_addons')
+                addons_url_list = get_gv_result_json(
+                    endpoint_url=configured_storage_addons_url,
+                    requesting_user=auth.user,
+                    requested_resource=project,
+                    request_method='GET',
+                    params={}
+                )
```
Comment (opaduchak) [inline website/citations/views.py:221]: you can use `get_addon` call for that.
It will return configured addon with external service included, whom which you can get `wb_key` and filter only requested addons

File: website/citations/views.py [hunk 9]
```diff
@@ -183,7 +196,87 @@ def citation_list(self):
...
+                # citation_list = list(
+                #     filter(lambda x: x['attributes']['external_service_name'] == addon_short_name, addons_url_list))
+
+                # resource_references_response = get_gv_result(
+                #     endpoint_url=USER_LIST_ENDPOINT,
+                #     requesting_user=auth.user,
+                #     params={'filter[user_uri]': f'{settings.DOMAIN}/{auth.user._id}'},
+                # )
+                for addon in addons_url_list:
```
Comment (opaduchak) [inline website/citations/views.py:230]: Nit: maybe do this concurrently?

File: website/citations/views.py [hunk 10]
```diff
@@ -183,7 +196,87 @@ def citation_list(self):
...
+                    gv_response = _make_gv_request(
+                        endpoint_url=f'{settings.GRAVYVALET_URL}/v1/addon-operation-invocations/',
+                        requesting_user=auth.user,
+                        requested_resource=project,
+                        request_method='POST',
+                        params={},
```
Comment (opaduchak) [inline website/citations/views.py:236]:
```suggestion
```

**Review by opaduchak — CHANGES_REQUESTED** [review]

File: addons/zotero/views.py [hunk 11]
```diff
@@ -57,12 +64,80 @@ def library_list(self):
...
+                            response.append({
+                                'addon': addon_short_name,
+                                'kind': item.get('item_type', '').lower(),
+                                'id': item_id,
+                                'name': item.get('item_name', ''),
+                                'path': item.get('item_path', ''),
+                                'parent_list_id': None,
+                                'provider_list_id': item_id,
+                            })
+                return response
```
Comment (opaduchak) [inline addons/zotero/views.py:137]: AFAIK this is only used during configuration of Zotero addon in legacy way, and therefore we should not need this.Please confirm whether we really need this, and if so, please reuse `citation_list_gv_request` function

File: osf/models/spam.py [hunk 12]
```diff
@@ -200,7 +200,6 @@ def do_check_spam(self, author, author_email, content, request_headers):
             'user_agent': request_headers.get('User-Agent'),
             'referer': request_headers.get('Referer'),
         }
-        request_kwargs.update(request_headers)
```
Comment (opaduchak) [inline osf/models/spam.py:203]: can you elaborate on this change, please?

**Review by opaduchak — CHANGES_REQUESTED** [review]

File: osf/models/spam.py (hunk 12 above)
Comment (Vlad0n20) [reply osf/models/spam.py:203]: It's not mine. It appeared when I merge develop branch

Comment (felliott) [reply osf/models/spam.py:203]: Yep. This was was a hotfix to avoid some breakage in some spam checks.  If this is still being applied, you may want to rebase against current develop.

**Comment by brianjgeiger** [comment]
@Vlad0n20 Could you fix the merge conflicts now that we are pointing to the `feature/gravy_valet_integration` branch?

**Comment by brianjgeiger** [comment]
@Vlad0n20 As mentioned in Open House, looks like there's a circular import that needs fixing. 

```
Testing modules "['addons']"
  File "/home/runner/work/osf.io/osf.io/osf/models/archive.py", line 10, in <module>
    from addons.base.models import BaseStorageAddon
  File "/home/runner/work/osf.io/osf.io/addons/base/models.py", line 15, in <module>
    from osf.models.node import AbstractNode
  File "/home/runner/work/osf.io/osf.io/osf/models/node.py", line 52, in <module>
    from .user import OSFUser
  File "/home/runner/work/osf.io/osf.io/osf/models/user.py", line 38, in <module>
    from osf.external.gravy_valet import (
  File "/home/runner/work/osf.io/osf.io/osf/external/gravy_valet/request_helpers.py", line 10, in <module>
    from osf.models import Node
ImportError: cannot import name 'Node' from partially initialized module 'osf.models' (most likely due to a circular import) (/home/runner/work/osf.io/osf.io/osf/models/__init__.py)
```

**Comment by brianjgeiger** [comment]
@Vlad0n20 Looks like you have some failing tests on the OSF side as well.

File: osf/models/node.py [hunk 13]
```diff
@@ -1191,20 +1202,19 @@ def callback(self, callback, recursive=False, *args, **kwargs):
 
     def can_comment(self, auth):
         if self.comment_level == 'public':
-            return auth.logged_in and (
-                self.is_public or
-                (auth.user and self.has_permission(auth.user, READ))
-            )
+            return auth.logged_in and self.is_public or (auth.user and self.has_permission(auth.user, READ))
```
Comment (brianjgeiger) [inline osf/models/node.py:1205]: Why did you change the parentheses here?

File: website/project/views/node.py [hunk 14]
```diff
@@ -711,6 +756,8 @@ def _view_project(node, auth, primary=False,
     view_only_link = auth.private_key or request.args.get('view_only', '').strip('/')
     anonymous = has_anonymous_link(node, auth)
     addons = list(node.get_addons())
+    for i in addons:
+        print(i.__dict__)
```
Comment (brianjgeiger) [inline website/project/views/node.py:760]: Do you still need this?

**Review by brianjgeiger — CHANGES_REQUESTED** [review]
A couple of small changes. Plus get the tests working, of course.

File: website/project/views/node.py (hunk 14 above)
Comment (Vlad0n20) [reply website/project/views/node.py]: Removed it

File: osf/models/node.py (hunk 13 above)
Comment (brianjgeiger) [reply osf/models/node.py:1205]: Oh? What kind of error? It's probably not a big deal, but it might be better to, for example, set a variable for `user_can_read = auth.user and self.has_permission(auth.user, READ)` then  shrink down the logic statement to work with that. It's likely logically equivalent, but the intent was a little more readable with the parentheses the way they were.

**Review by brianjgeiger — CHANGES_REQUESTED** [review]
I think the change in parentheses on the logic statement is actually consequential, so please make the changes I suggest there to simplify the statement. Other than that, this is ready to merge (once the tests pass on https://github.com/CenterForOpenScience/gravyvalet/pull/117 )

File: website/project/views/node.py [hunk 15]
```diff
@@ -480,17 +513,28 @@ def view_project(auth, node, **kwargs):
...
-        node_addon = node.get_addon('zotero')
-        zotero_widget_data = ZoteroCitationsProvider().widget(node_addon)
-        addons_widget_data['zotero'] = zotero_widget_data
-
-    if 'mendeley' in ret['addons']:
-        node_addon = node.get_addon('mendeley')
-        mendeley_widget_data = MendeleyCitationsProvider().widget(node_addon)
-        addons_widget_data['mendeley'] = mendeley_widget_data
-
+    if not waffle.flag_is_active(request, features.ENABLE_GV):
```
Comment (opaduchak) [inline website/project/views/node.py:516]: nit:
it's better for readability to flip this if statement

File: website/citations/views.py [hunk 16]
```diff
@@ -183,7 +193,19 @@ def citation_list(self):
         def _citation_list(auth, node_addon, list_id=None, **kwargs):
             """ Returns a list of citations
             """
+
             show = request.args.get('view', 'all')
-            return Provider().citation_list(node_addon, auth.user, list_id, show)
+            if not waffle.flag_is_active(request, features.ENABLE_GV):
```
Comment (opaduchak) [inline website/citations/views.py:198]: nit:
flip if statement

File: osf/models/node.py (hunk 13 above)
Comment (opaduchak) [reply osf/models/node.py:1205]: Parenthesis here are not for formatting, this changes logic and therefore
image

File: osf/external/gravy_valet/translations.py [hunk 17]
```diff
@@ -65,6 +77,11 @@ class EphemeralAddonConfig:
     label: str
     short_name: str
     full_name: str
+    has_widget: bool = False
```
Comment (opaduchak) [inline osf/external/gravy_valet/translations.py:80]: I'd convert this to a property or exclude this from init
```suggestion
    has_widget: bool = dataclasses.field(init=False, default=False)
```

File: osf/external/gravy_valet/request_helpers.py [hunk 18]
```diff
@@ -154,13 +192,159 @@ def _make_gv_request(
...
+                changed_response.append({
+                    'addon': addon_short_name,
+                    'kind': item.get('item_type', '').lower(),
+                    'id': item_id,
+                    'name': item.get('item_name', ''),
+                    'path': item.get('item_path', ''),
+                    'parent_list_id': None,
+                    'provider_list_id': item_id,
+                })
+    return changed_response
```
Comment (opaduchak) [inline osf/external/gravy_valet/request_helpers.py:345]: This request is no longer needed when `gravy_waffle` is enabled, but I'd advice you to test this out before making changes

File: osf/external/gravy_valet/request_helpers.py [hunk 19]
```diff
@@ -154,13 +192,159 @@ def _make_gv_request(
...
+    gv_addon_name = addon_short_name if addon_short_name != 'zotero' else 'zotero_org'
+    citation_url_list = list(
+        filter(lambda x: x['attributes']['external_service_name'] == gv_addon_name, addons_url_list))
+    return citation_url_list
+
+
+def _get_gv_response(auth, addon, project, list_id):
+    data = {
+        'attributes': {
+            'operation_name': 'list_root_collections',
```
Comment (opaduchak) [inline osf/external/gravy_valet/request_helpers.py:237]: Can you elaborate on the reason why one may list root collections only?

File: osf/external/gravy_valet/request_helpers.py [hunk 20]
```diff
@@ -154,13 +192,159 @@ def _make_gv_request(
...
+        request_method='GET',
+        params={}
+    )
+    gv_addon_name = addon_short_name if addon_short_name != 'zotero' else 'zotero_org'
+    citation_url_list = list(
+        filter(lambda x: x['attributes']['external_service_name'] == gv_addon_name, addons_url_list))
+    return citation_url_list
+
+
+def _get_gv_response(auth, addon, project, list_id):
```
Comment (opaduchak) [inline osf/external/gravy_valet/request_helpers.py:234]: this has too generic function name, maybe rename it?

File: osf/external/gravy_valet/request_helpers.py [hunk 21]
```diff
@@ -154,13 +192,159 @@ def _make_gv_request(
...
+        response = requests.post(endpoint_url, headers=auth_headers, json={'data': data})
+    else:
+        response = requests.get(endpoint_url, headers=auth_headers, params=params)
     if not response.ok:
         # log error to Sentry
         pass
     return response
 
 
+def get_citation_url_list(auth, addon_short_name, project, request=None, pid=None):
```
Comment (opaduchak) [inline osf/external/gravy_valet/request_helpers.py:206]: I'd rename this method, as it's name isn't clear

File: osf/external/gravy_valet/request_helpers.py [hunk 22]
```diff
@@ -114,12 +132,32 @@ def get_gv_result(
...
+    response_json = _make_gv_request(
+        endpoint_url=endpoint_url,
+        requesting_user=requesting_user,
+        requested_resource=requested_resource,
+        request_method=request_method,
+        params=params,
+    ).json()
+    if not response_json.get('data'):
+        return dict()
+    return response_json['data']
```
Comment (opaduchak) [inline osf/external/gravy_valet/request_helpers.py:152]: I'd name it `get_raw_gv_result`

File: addons/zotero/views.py [hunk 23]
```diff
@@ -57,12 +61,19 @@ def library_list(self):
         def _library_list(auth, node_addon, **kwargs):
             """ Returns a list of group libraries - for use with Zotero addon
             """
+
             limit = request.args.get('limit')
             start = request.args.get('start')
             return_count = is_truthy(request.args.get('return_count', False))
             append_personal = is_truthy(request.args.get('append_personal', True))
-            return node_addon.get_folders(limit=limit, start=start, return_count=return_count, append_personal=append_personal)
+            if not waffle.flag_is_active(request, features.ENABLE_GV):
```
Comment (opaduchak) [inline addons/zotero/views.py:69]: nit: I'd also flip this statement

Comment (opaduchak) [reply addons/zotero/views.py:69]: Moreover, there is a change that we don't need lib library list anymore

**Review by opaduchak — CHANGES_REQUESTED** [review]
Great work!
I've added some suggestions and parenthesis thing definitely needs to be change

**Comment by brianjgeiger** [comment]
Also, there are now some merge conflicts to be resolved.

File: osf/external/gravy_valet/request_helpers.py (hunk 19 above)
Comment (Vlad0n20) [reply osf/external/gravy_valet/request_helpers.py:237]: In 250 row I check if list_id exists and change to list_collection_items

File: addons/zotero/views.py (hunk 23 above)
Comment (Vlad0n20) [reply addons/zotero/views.py]: I tried to remove library_list but too many errors appeared

File: osf/external/gravy_valet/request_helpers.py (hunk 21 above)
Comment (Vlad0n20) [reply osf/external/gravy_valet/request_helpers.py]: done

File: osf/external/gravy_valet/request_helpers.py (hunk 20 above)
Comment (Vlad0n20) [reply osf/external/gravy_valet/request_helpers.py]: Renamed

File: osf/external/gravy_valet/request_helpers.py (hunk 18 above)
Comment (Vlad0n20) [reply osf/external/gravy_valet/request_helpers.py]: I tried to remove library_list but too many errors appeared

File: osf/external/gravy_valet/translations.py (hunk 17 above)
Comment (Vlad0n20) [reply osf/external/gravy_valet/translations.py]: Done

File: website/citations/views.py (hunk 16 above)
Comment (Vlad0n20) [reply website/citations/views.py]: Done

File: website/project/views/node.py (hunk 15 above)
Comment (Vlad0n20) [reply website/project/views/node.py]: Done

File: osf/models/node.py (hunk 13 above)
Comment (Vlad0n20) [reply osf/models/node.py]: Return to previous version

File: osf/external/gravy_valet/request_helpers.py (hunk 19 above)
Comment (opaduchak) [reply osf/external/gravy_valet/request_helpers.py:251]: Can you do this either on 237 line please, or 264, but not on both, because this is confusing

File: osf/external/gravy_valet/request_helpers.py (hunk 18 above)
Comment (opaduchak) [reply osf/external/gravy_valet/request_helpers.py]: Have you tried removing GV library list only? 
And may you showcase which errors have appeared, please
?

File: osf/external/gravy_valet/request_helpers.py [hunk 24]
```diff
@@ -179,6 +217,114 @@ def _make_gv_request(
...
+        request_method='GET',
+        params={}
+    )
+    gv_addon_name = addon_short_name if addon_short_name != 'zotero' else 'zotero_org'
+    citation_url_list = list(
+        filter(lambda x: x['attributes']['external_service_name'] == gv_addon_name, addons_url_list))
+    return citation_url_list
+
+
+def _get_gv_addon_operation_invocations_response(auth, addon, project, list_id):
```
Comment (opaduchak) [inline osf/external/gravy_valet/request_helpers.py:248]: this is still not clear enough, because this invokes only citation operations

**Review by opaduchak — CHANGES_REQUESTED** [review]
please rename one more function and we're good to go

File: osf/external/gravy_valet/request_helpers.py (hunk 24 above)
Comment (Vlad0n20) [reply osf/external/gravy_valet/request_helpers.py]: Done

**Comment by brianjgeiger** [comment]
@opaduchak Are these changes sufficient?

**Comment by brianjgeiger** [comment]
@Vlad0n20 There's now a merge conflict due to a recent bug fix. Sorry about that.

File: osf/external/gravy_valet/request_helpers.py (hunk 19 above)
Comment (opaduchak) [reply osf/external/gravy_valet/request_helpers.py:251]: @Vlad0n20 
It'd be great if you fixed that one

**Review by opaduchak — APPROVED** [review]

**Comment by brianjgeiger** [comment]
Looks like there's one failing test:
```
=========================== short test summary info ============================
FAILED osf_tests/test_gv_utils.py::TestRequestHelpers::test_get_user_accounts - AttributeError: '_FakeAccount' object has no attribute 'base_account'
=== 1 failed, 3334 passed, 29 skipped, 23823 warnings in 1430.70s (0:23:50) ====
```
